File: chapters/chp7_code/Log.c
#include <stdio.h>
#include <math.h>

void generate_log_surface(double r_min, double r_max, double theta_min, double theta_max,
                         int nr, int ntheta, const char* filename) {
    FILE* file = fopen(filename, "w");
    if (!file) return;
    
    fprintf(file, "x,y,z,real_log,imag_log\n");
    
    double dr = (r_max - r_min) / nr;
    double dtheta = (theta_max - theta_min) / ntheta;
    
    for (int i = 0; i <= nr; i++) {
        for (int j = 0; j <= ntheta; j++) {
            double r = r_min + i * dr;
            double theta = theta_min + j * dtheta;
            
            double x = r * cos(theta);
            double y = r * sin(theta);
            double z = theta;
            
            double real_log = log(r);
            double imag_log = theta;
            
            fprintf(file, "%.6f,%.6f,%.6f,%.6f,%.6f\n",
                    x, y, z, real_log, imag_log);
        }
    }
    
    fclose(file);
    printf("Riemann surface data saved to %s\n", filename);
}

int main() {
    printf("Riemann Surface for log(z) (C version)\n");
    printf("======================================\n");
    
    generate_log_surface(0.1, 2, -4*M_PI, 4*M_PI, 100, 200, "riemann_log_c.csv");
    
    return 0;
}

File: chapters/chp7_code/Log.rb
#!/usr/bin/env ruby

class RiemannSurface
  def self.generate_log_surface(r_range, theta_range, nr, ntheta, filename)
    File.open(filename, 'w') do |file|
      file.puts "x,y,z,real_log,imag_log"
      
      r_min, r_max = r_range
      theta_min, theta_max = theta_range
      
      dr = (r_max - r_min) / nr.to_f
      dtheta = (theta_max - theta_min) / ntheta.to_f
      
      (0..nr).each do |i|
        (0..ntheta).each do |j|
          r = r_min + i * dr
          theta = theta_min + j * dtheta
          
          x = r * Math.cos(theta)
          y = r * Math.sin(theta)
          z = theta
          
          # Complex logarithm: log(z) = ln|z| + i*arg(z)
          complex_z = Complex(x, y)
          log_z = Complex.log(complex_z)
          
          file.puts "#{x},#{y},#{z},#{log_z.real},#{log_z.imag}"
        end
      end
    end
    
    puts "Riemann surface data saved to #{filename}"
  end
end

puts "Riemann Surface for log(z) (Ruby version)"
puts "=========================================="

RiemannSurface.generate_log_surface([0.1, 2], [-4*Math::PI, 4*Math::PI], 100, 200, "riemann_log_ruby.csv")